code/datums/components/rust.dm

```diff
@@ -0,0 +1,83 @@
+/datum/component/rust
+	dupe_mode = COMPONENT_DUPE_HIGHLANDER
+	var/atom/parent_atom
+	var/mutable_appearance/rust_overlay
+	var/static/list/typecache_of_valid_types = list(
+		/turf/closed/wall,
+		/turf/closed/wall/r_wall,
+		/turf/open/floor/plating,
+	)
+
+/turf/closed/wall/rust/New()
+	var/atom/T = new /turf/closed/wall(src)
+	T._AddComponent(list(/datum/component/rust))
+
+/turf/closed/wall/r_wall/rust/New()
+	var/atom/T = new /turf/closed/wall/r_wall(src)
+	T._AddComponent(list(/datum/component/rust))
+
+/turf/open/floor/plating/rust/New()
+	var/atom/T = new /turf/open/floor/plating(src)
+	T._AddComponent(list(/datum/component/rust))
+
+/datum/component/rust/Initialize(...)
+	. = ..()
+	if(!(parent.type in typecache_of_valid_types) || !isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	parent_atom = parent
+	if(!("rust" in icon_states(parent_atom.icon)) || parent_atom.GetExactComponent(/datum/component/rust))
```

Comment:
Icon state should just be passed in as an argument.

Also, just set `dupe_mode` to `COMPONENT_DUPE_HIGHLANDER`.